src/Microsoft.ML.Recommender/doc.xml

```diff
@@ -0,0 +1,90 @@
+        // The variable _synthesizedMatrixFirstRowIndex indicates the integer that would be mapped to the first row index. If user data uses
+        // 0-based indices for rows, _synthesizedMatrixFirstRowIndex can be set to 0. Similarly, for 1-based indices, _synthesizedMatrixFirstRowIndex
+        // could be 1.
+        const int _synthesizedMatrixFirstColumnIndex = 1;
+        const int _synthesizedMatrixFirstRowIndex = 1;
+        const int _synthesizedMatrixColumnCount = 60;
+        const int _synthesizedMatrixRowCount = 100;
+
+        class MatrixElement
+        {
+            // Matrix column index starts from 1 and is at most _synthesizedMatrixColumnCount.
+            // Contieuous=true means that all values from 1 to _synthesizedMatrixColumnCount are allowed keys.
+            [KeyType(Contiguous = true, Count = _synthesizedMatrixColumnCount, Min = _synthesizedMatrixFirstColumnIndex)]
+            public uint MatrixColumnIndex;
+            // Matrix row index starts from 1 and is at most _synthesizedMatrixRowCount.
+            // Contieuous=true means that all values from 1 to _synthesizedMatrixRowCount are allowed keys.
+            [KeyType(Contiguous = true, Count = _synthesizedMatrixRowCount, Min = _synthesizedMatrixFirstRowIndex)]
+            public uint MatrixRowIndex;
+            // The value at the MatrixColumnIndex-th column and the MatrixRowIndex-th row in the considered matrix.
+            public float Value;
+        }
+
+        // Create an in-memory matrix as a list of tuples (column index, row index, value).
+        var dataMatrix = new List<MatrixElement>();
+        for (uint i = _synthesizedMatrixFirstColumnIndex; i < _synthesizedMatrixFirstColumnIndex + _synthesizedMatrixColumnCount; ++i)
+            for (uint j = _synthesizedMatrixFirstRowIndex; j < _synthesizedMatrixFirstRowIndex + _synthesizedMatrixRowCount; ++j)
+                dataMatrix.Add(new MatrixElement() { MatrixColumnIndex = i, MatrixRowIndex = j, Value = (i + j) % 5 });
+
+        // Convert the in-memory matrix into an IDataView so that ML.NET components can consume it.
+        var dataView = ComponentCreation.CreateDataView(Env, dataMatrix);
+
+        // Create a matrix factorization trainer which may consume "Value" as the training label, "MatrixColumnIndex" as the
+        // matrix's column index, and "MatrixRowIndex" as the matrix's row index.
+        var mlContext = new MLContext(seed: 1, conc: 1);
+        var pipeline = new MatrixFactorizationTrainer(mlContext, "Value", "MatrixColumnIndex", "MatrixRowIndex",
+            advancedSettings: s =>
+            {
+                s.NumIterations = 10;
+               s.NumThreads = 1; // To eliminate randomness, # of threads must be 1.
+                s.K = 32;
+            });
+
+        // Train a matrix factorization model.
+        var model = pipeline.Fit(dataView);
```

Comment:
please move this in the Samples project, on its own file. See: 

[https://github.com/dotnet/machinelearning/blob/master/docs/samples/Microsoft.ML.Samples/Static/Trainers.cs#L22](https://github.com/dotnet/machinelearning/blob/master/docs/samples/Microsoft.ML.Samples/Static/Trainers.cs#L22) 

and 
[https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.StandardLearners/Standard/SdcaStatic.cs#L41](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.StandardLearners/Standard/SdcaStatic.cs#L41)

Reply:
The test is not a complete example.

Reply:
Sorry I don't understand. Could you tell me which part is missing?

---
In reply to: [229474507](https://github.com/dotnet/machinelearning/pull/1450#discussion_r229474507) [](ancestors = 229474507)